ocean-client/src/programs/lm-reinvest-program.ts

```diff
@@ -0,0 +1,159 @@
+        if (!this.settings.reinvestThreshold || this.settings.reinvestThreshold <= 0) {
+            return false
+        }
+
+        const utxoBalance = await this.getUTXOBalance()
+        const tokenBalance = balances.get("DFI")
+
+        const amountFromBalance = new BigNumber(tokenBalance?.amount ?? "0")
+        const fromUtxos = utxoBalance.gt(1) ? utxoBalance.minus(1) : new BigNumber(0)
+        let amountToUse = fromUtxos.plus(amountFromBalance)
+
+        let prevout: Prevout | undefined = undefined
+        console.log("checking for reinvest: " + fromUtxos + " from UTXOs, " + amountFromBalance + " tokens. total " + amountToUse + " vs " + this.settings.reinvestThreshold)
+        if (amountToUse.gt(this.settings.reinvestThreshold) && fromUtxos.gt(0)) {
+            console.log("converting " + fromUtxos + " UTXOs to token ")
+            const tx = await this.utxoToOwnAccount(fromUtxos)
+            prevout = this.prevOutFromTx(tx)
+        }
+
+        if (amountToUse.gt(this.settings.reinvestThreshold)) {
+            let donatedAmount = new BigNumber(0);
+            if (this.settings.autoDonationPercentOfReinvest > 0) {
+                //send donation and reduce amountToUse
+                donatedAmount = amountToUse.times(this.settings.autoDonationPercentOfReinvest).div(100)
+                console.log("donating " + donatedAmount.toFixed(2) + " DFI")
+                const tx = await this.sendDFIToAccount(donatedAmount, DONATION_ADDRESS, prevout)
+                prevout = this.prevOutFromTx(tx)
+
+                amountToUse = amountToUse.minus(donatedAmount)
+            }
+
+            const amountToSwap = amountToUse.div(2) //half is swapped to other token
+            const tokenA = pool.tokenA
+            const tokenB = pool.tokenB
+            console.log("swaping " + amountToSwap + " half of (" + amountFromBalance + "+" + fromUtxos + "-" + donatedAmount + ") DFI to " + tokenA.symbol)
+            const swap = await this.swap(amountToUse, 0, +tokenA.id, new BigNumber(999999999), prevout)
+            if (! await this.waitForTx(swap.txId)) {
+                await telegram.send("ERROR: swapping reinvestment failed")
+                console.error("swapping reinvestment failed")
+                return false
+            }
+
+            const updatedPool = await this.getPool(this.lmPair)
+            pool = updatedPool!
+            const tokens = await this.getTokenBalances()
+            const availableA = new BigNumber(tokens.get(tokenA.symbol)?.amount ?? 0)
+            let usedAssetB = new BigNumber(tokens.get(tokenB.symbol)?.amount ?? 0)
+            let usedAssetA = usedAssetB.multipliedBy(pool.priceRatio.ab)
+            if (usedAssetA.gt(availableA)) { //not enough stocks to fill it -> use full stocks and reduce DUSD
+                usedAssetA = availableA
+                usedAssetB = usedAssetA.multipliedBy(pool.priceRatio.ba)
+            }
```

Comment:
this code block is just sanity checking if we have enough to be added to the pool and do a short correction, if pool has changed heavily?

Reply:
yes exactly. after the swap, we might not have perfect ratio anymore. don't want to loose assets due to addLiquidity-slipage ;)